### ISSUE TITLE
Lint and svg flag

### ISSUE THREAD

**Issue by TalbotG** [web-platform-tests/wpt#23253]
@foolip Philip,

I presume you are one of the wpt maintainers.
Is the svg flag valid or deprecated? 
`<meta content="svg" name="flags">`
I have declared the svg flag in 2 tests
and in 2 reference files so far:
https://github.com/web-platform-tests/wpt/pull/20033
https://github.com/web-platform-tests/wpt/pull/21450
**svg** is not in the list of **valid flags**
http://web-platform-tests.org/writing-tests/css-metadata.html#requirement-flags
nor in the list of **deprecated flags** . If
svg flag is incorrect or deprecated, then why the lint verification did not report it?

**Comment by foolip** [comment]
@TalbotG I am indeed still active in WPT, but to get the attention of a wider group of people maintaining it, you can mention @web-platform-tests/wpt-core-team.

It looks like the problem is that there are no lints for any flags at all, valid or invalid. I've renamed the issue to reflect this.

Is the "svg" flag useful for you to add to tests? If so I'd say you can simply keep doing so, and perhaps add a description of when it's useful in http://web-platform-tests.org/writing-tests/css-metadata.html#requirement-flags.

**Comment by TalbotG** [comment]
> Is the "svg" flag useful for you to add to tests?

I used to add the "svg" flag for tests that were using an svg object. This was the normal procedure before.

"svg | Requires support for vector graphics (SVG)"
https://wiki.csswg.org/test/format#requirement-flags

There are some web-capable softwares that want to be and/or strive to become (now or in the future) W3C-web-standards-compliant (at least regarding CSS) which may not have support for SVG objects. Therefore why the usage of the svg flag arised. Let's say they import, set up and run the entire CSS repository: they do not want to run tests that have svg objects because their software can not handle those. Then such flag would be useful to them.

I would say that flags' items are useful for filtering or screening tests (regarding a specific criteria) for those who import, create and run repositories of tests.

**Comment by TalbotG** [comment]
> to get the attention of a wider group of people maintaining it, you can mention @web-platform-tests/wpt-core-team 

I wrote down this (useful!) information. Thank you Philip

**Comment by stephenmcgruer** [comment]
If this is a flag that the css folks feel is useful, I think that documenting it (the file is `docs/writing-tests/css-metadata.md, PRs are welcome ;)) is a good idea.

Additionally, a lint for `css/` against deprecated flags seems reasonable too. 

I'm not clear what other lints would be both practical and useful. Meta tags are author-defined, and some authors may with to use them with different `name`s or `content`s and so I'm not sure we should impose there (even for cases like the flag typos @foolip found). Open to ideas though!

**Comment by TalbotG** [comment]
> If this is a flag that the css folks feel is useful,

I do not know. 
I have checked the documentation of some web-aware and CSS-capable user agents like [Prince 13](https://www.princexml.com/doc/graphics/#scalable-vector-graphics-svg) and [Antenna House 7](https://www.antenna.co.jp/AHF/help/v70e/ahf-gra.html#SVG) and to my surprise, they appear to support (at least partially) SVG objects. 
Some other newer web-aware and CSS-capable user agents (eg [gngr](https://gngr.info/) ) do not support SVG objects.

> a lint for css/ against deprecated flags seems reasonable too.

I am for a lint for css/ checking against deprecated flags and also unknown or invalid flags. Eg https://github.com/web-platform-tests/wpt/pull/23264 
"border-radius" and "box-shadow" are not - and never were - valid flags.

**Comment by TalbotG** [comment]
> Some other newer web-aware and CSS-capable user agents (eg gngr ) do not support SVG objects.

With more research, even gngr is looking, planning to support svg objects:
[Support for SVG images in gngr](https://github.com/UprootLabs/gngr/issues/210)

**Comment by prsmahajan** [comment]
Hi, I would like to work on this issue, Can I get started with this?

**Comment by foolip** [comment]
@prsmahajan absolutely, a fix for this would be welcome!

The existing lints are in https://github.com/web-platform-tests/wpt/blob/master/tools/lint/lint.py and I think this would be a matter of adding something to this method:
https://github.com/web-platform-tests/wpt/blob/198dd874ceede59b072bd650a027a3bf619f94ff/tools/lint/lint.py#L494

Basically, find `<meta name="flags">` elements and then check the `contents` attribute of those.

Please comment here if you need more help.

**Comment by prsmahajan** [comment]
Do you need to insert more meta tags in this file?

**Comment by foolip** [comment]
@prsmahajan no, the linter only checks the content, it doesn't make any modifications.

**Comment by TalbotG** [comment]
@prsmahajan ,
The list of valid flags are in
http://web-platform-tests.org/writing-tests/css-metadata.html#requirement-flags
and, in my opinion, such list should include svg .

**Comment by prsmahajan** [comment]
At which line do you need to modify?

**Comment by foolip** [comment]
@prsmahajan I don't know exactly without writing the change myself. The order isn't super important, but what you'll want to add will be similar to https://github.com/web-platform-tests/wpt/blob/198dd874ceede59b072bd650a027a3bf619f94ff/tools/lint/lint.py#L587.

**Comment by prsmahajan** [comment]
I'll have a look at it and let you know Senior :)

**Comment by DanielRyanSmith** [comment]
Hi, I'm interested in taking on this issue if it's still available.

I have some clarifying questions about the issue:
- Are the list of requirement flags up to date in the docs [here](http://web-platform-tests.org/writing-tests/css-metadata.html#requirement-flags)? Will part of this issue be to add `svg` as a valid flag?
- Is there a good place in the repository to reference the list of valid and deprecated CSS requirement flags to compare in the lint file? Is it worth creating a constants file in the lint directory to do this?
- Just to make sure, this lint will only apply to the files in the `css/` directory, yes?
- From what I can see already, there will be a number of cases where running the lint will now pick up invalid/deprecated flags. Will these need to be fixed at the same time or will fixing these be a separate issue?

**Comment by TalbotG** [comment]
>     * Are the list of requirement flags up to date in the docs http://web-platform-tests.org/writing-tests/css-metadata.html#requirement-flags ? 

In my opinion, no, such list of requirement flags is not perfectly up to date. In my opinion, the `svg` flag should be in the list of valid flags. 
{
Addendum: [Geoffrey Sneddon in August 2016](https://lists.w3.org/Archives/Public/public-css-testsuite/2016Aug/0016.html
) wrote 
"
I believe we have rough consensus to drop:

* ahem
* http
* image
* namespace
* svg
* 32bit
* 96dpi
(...)

image, namespace, and svg are all catering for theoretical UAs that
don't implement the respective features.
"
So maybe I am wrong on this `svg` issue.
}

>  Will part of this issue be adding `svg` as a valid flag?

Yes.
 
>     * Is there a good place in the repository to reference the list of valid and deprecated CSS requirement flags to compare in the lint file? 

I believe that the best place in the repository for such reference should be the current file: 
http://web-platform-tests.org/writing-tests/css-metadata.html#requirement-flags

>     * Will this lint check only apply to the files in the `css/` directory only or any file with `<meta name="flags">` elements?

I believe the lint check should only apply to the files in the `css/` directory only. But, to be extra sure, maybe you should ask @gsnedders 

>     * From what I can see already, there will be a number of cases where running the full lint will now pick up invalid/deprecated flags. Will these need to be fixed at the same time or will fixing these be a separate issue?

I can not answer you on this. I guess fixing these could be a separate issue. What are the possible and reasonable consequences of picking up css tests with deprecated flags? Running the full lint on the test suite may be a risky adventure... I am **not confident** in how to reply to your question. I do not know the possible and reasonable consequences of picking up css tests with deprecated flags.

> these are the requirement flags I am seeing as invalid/deprecated and the number of times I see them with my initial attempt at this:
> 
> ```
> {
> 	'ahem': 2966,
> 	'image': 1688,
> 	'dom': 883,
> 	'svg': 700,
> 	'interact': 666,
> 	'font': 223,
> 	'animated': 125,
> 	'namespace': 124,
> 	'32bit': 65,
> 	'http': 55,
> 	'Ahem': 41,
> 	'userstyle': 28,
> 	'96dpi': 14,
> 	'history': 7,
> 	'speech': 7,
> 	'combo': 2,
> 	'internal': 1,
> 	'may21': 1,
> 	'page': 1,
> }
> ```

In your list above, 'Ahem': 41, 'page': 1, 'may21': 1 and 'internal': 1 are incorrect, invalid flags and were not even in the original list:
https://wiki.csswg.org/test/format#requirement-flags

The others you listed above became deprecated flags... except `svg`.
I think you need to check with @gsnedders or maybe @stephenmcgruer what could happen or will happen if you remove all of the deprecated flags, including 'animated' and 'interact' because 
"Tests which satisfy the described criteria should simply be designated as “manual” using the -manual file name flag."
coming from
http://web-platform-tests.org/writing-tests/css-metadata.html#requirement-flags 

But for me, you can safely remove 'Ahem', 'page', 'may21' and 'internal' from css tests because those are incorrect, invalid.

For reference, we started to have discussion in the css-testsuite mailing list in this August 2016 thread:

https://lists.w3.org/Archives/Public/public-css-testsuite/2016Aug/0016.html

**Comment by foolip** [comment]
> From what I can see already, there will be a number of cases where running the full lint will now pick up invalid/deprecated flags. Will these need to be fixed at the same time or will fixing these be a separate issue?

There are two ways. Either fix all of the lint errors before landing the lint. Or, suppress the lint failures in `lint.ignore` and then fix them in smaller batches after the lint has been enabled.

**Comment by shreyamaurya029** [comment]
is this beginner friendly??
if yes, can u assign me this issue i would like to work on it

**Comment by DanielRyanSmith** [comment]
> I don't think the requirements are clear enough to make this a beginner friendly issue, but maybe others disagree.

Agreed. I took a shot at this issue a while ago and there's a bit of an iceberg here in terms of work (some deprecated flags still function under the hood and require file renaming before removal, all unused flags need to be removed before adding the lint check, etc). Some parts of this issue I began addressing, but it became clear that the marginal value of adding this functionality outweighed the work required to make it happen. Thanks for the offer @shreyamaurya029, but I think this is incorrectly marked as a good first issue and time is best spent on other tasks.
